### ISSUE TITLE
x/tools/gopls: support "go generate" via CodeLens

### ISSUE THREAD

**Issue by marwan-at-work** [golang/go#37680]
**TL;DR:** 

It would be helpful if gopls recognized `//go:generate` directives in Go files and offered the option to run "go generate" from the editor through a CodeLens. Here's a demonstration video: https://vimeo.com/395630819

**Motivation** - feel free to skip if you are already convinced :) 
Code generation is widely used in Go projects. The `//go:generate` command makes long and tedious commands much easier to execute by just running `go generate` or `go generate ./...`. 

However, whenever a programmer wants to regenerate the code, they end up doing some context switching from their editor to the terminal to regenerate the code. 

Therefore, if the editor understood the `//go:generate` detective and offered to run `go generate` on the user's behalf, then the programmer can remain within the editor and not worry about switching contexts. 

VSCode already does this with Go Test functions by giving you the option to run tests from within the UI. The "go generate" feature was actually requested for a few years in the VSCode extension but was never implemented: https://github.com/microsoft/vscode-go/issues/263

Therefore, I think gopls is the right place to offer this functionality as the feature is very Go-specific. 

**Implementation question**

The video above is from an initial implementation I wrote here: https://github.com/golang/tools/compare/master...marwan-at-work:gg?expand=1

The CodeLens part seems super straightforward. However, the moment we click on "go generate" and the client sends the command back to gopls, we need to show the stdout/stderr of "go generate" back to the client. 

The one supported way I saw was using the `$/progress` server->client rpc which lets you show one message at a time. Each message shows as a progress bar on the bottom right in VSCode. However, would it be more appropriate to show this as a window instead of a progress bar?

If so, does the LSP provide a way to tell the editor to open a window and show a stream of logs?

In any case, if everyone is happy with this proposal, I'm happy to add tests and contribute the implementation above!

**Comment by stamblerre** [comment]
Completely agree with the idea. We've had a similar issue opened here: https://github.com/golang/go/issues/31817. I think the code lenses are a great starting point. 

I'm not sure if LSP has that functionality, as we haven't used any of the in-progress RPC features. Let's experiment with what's currently available, and if something seems to be missing, we can open issues with the LSP specification.

Thanks for doing this work, @marwan-at-work! Looking forward to getting these features in `gopls`!
